incubator/binding-asyncapi/src/main/java/io/aklivity/zilla/runtime/binding/asyncapi/internal/AsyncapiClientCompositeBindingAdapter.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021-2023 Aklivity Inc
+ *
+ * Licensed under the Aklivity Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ *   https://www.aklivity.io/aklivity-community-license/
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package io.aklivity.zilla.runtime.binding.asyncapi.internal;
+
+import static io.aklivity.zilla.runtime.engine.config.KindConfig.CLIENT;
+
+import io.aklivity.zilla.runtime.binding.asyncapi.config.AsyncapiConfig;
+import io.aklivity.zilla.runtime.binding.asyncapi.config.AsyncapiOptionsConfig;
+import io.aklivity.zilla.runtime.binding.tcp.config.TcpOptionsConfig;
+import io.aklivity.zilla.runtime.binding.tls.config.TlsOptionsConfig;
+import io.aklivity.zilla.runtime.engine.config.BindingConfig;
+import io.aklivity.zilla.runtime.engine.config.CompositeBindingAdapterSpi;
+import io.aklivity.zilla.runtime.engine.config.NamespaceConfigBuilder;
+
+public class AsyncapiClientCompositeBindingAdapter extends AsyncapiCompositeBindingAdapter implements CompositeBindingAdapterSpi
+{
+    private int[] mqttPorts;
+    private int[] mqttsPorts;
+
+    @Override
+    public String type()
+    {
+        return AsyncapiBinding.NAME;
+    }
+
+    @Override
+    public BindingConfig adapt(
+        BindingConfig binding)
+    {
+        AsyncapiOptionsConfig options = (AsyncapiOptionsConfig) binding.options;
+        AsyncapiConfig asyncapiConfig = options.specs.get(0);
+        this.asyncApi = asyncapiConfig.asyncApi;
+
+        this.allPorts = resolveAllPorts();
+        this.mqttPorts = resolvePortsForScheme("mqtt");
+        this.mqttsPorts = resolvePortsForScheme("mqtts");
+        this.isPlainEnabled = mqttPorts != null;
+        this.isTlsEnabled = mqttsPorts != null;
+        this.qname = binding.qname;
```

Comment:
Please avoid computing state we don't need and move fields to server only if that's where they are needed.